src/librustc_mir/transform/nrvo.rs

```diff
@@ -0,0 +1,225 @@
+use rustc_index::bit_set::HybridBitSet;
+use rustc_middle::mir::visit::{MutVisitor, PlaceContext, Visitor};
+use rustc_middle::mir::{self, BasicBlock, Local, Location};
+use rustc_middle::ty::TyCtxt;
+
+use crate::transform::{MirPass, MirSource};
+
+/// This pass looks for MIR that always copies the same local into the return place and eliminates
+/// the copy by renaming all uses of that local to `_0`.
+///
+/// This allows LLVM to perform an optimization similar to the named return value optimization
+/// (NRVO) that is guaranteed in C++. This avoids a stack allocation and `memcpy` for the
+/// relatively common pattern of allocating a buffer on the stack, mutating it, and returning it by
+/// value like so:
+///
+/// ```rust
+/// fn foo(init: fn(&mut [u8; 1024])) -> [u8; 1024] {
+///     let mut buf = [0; 1024];
+///     init(&mut buf);
+///     buf
+/// }
+/// ```
+///
+/// For now, this pass is very simple and only capable of eliminating a single copy.
+/// A more general version of copy propagation could yield even more benefits.
```

Comment:
```suggestion
/// A more general version for merging locals could yield even more benefits.
```

pure copy prop isn't as powerful as merging locals that never needed to be separate.

Reply:
In the literature I've read, "copy propagation" is a general idea, not a specific algorithm. If you have a counter-example, I would be happy to see it, although I feel quite strongly that my interpretation in correct. This PR, much like the existing `copy_prop` pass in `rustc`, is a very limited form of copy propagation. The "place unification" algorithm in #47954 and #71003 is a more robust approach, and I will change this comment to link to it. Unlike copy propagation, "place unification", which I've been calling it "variable coalescing" because of the similar step during [register allocation](https://en.wikipedia.org/wiki/Register_allocation#Coalescing), is not a term of art.